tests/integration/conftest.py

```diff
@@ -0,0 +1,111 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Fixtures for Mattermost charm integration tests."""
+
+import asyncio
+import logging
+from pathlib import Path
+from urllib.parse import urlparse
+
+import pytest_asyncio
+import yaml
+from ops.model import ActiveStatus
+from pytest import fixture
+from pytest_operator.plugin import OpsTest
+
+logger = logging.getLogger(__name__)
+
+
+@fixture(scope="module", name="metadata")
+def metadata_fixture():
+    """Provides charm metadata."""
+    yield yaml.safe_load(Path("./metadata.yaml").read_text("utf-8"))
+
+
+@fixture(scope="module", name="app_name")
+def app_name_fixture(metadata):
+    """Provides app name from the metadata."""
+    yield metadata["name"]
+
+
+@fixture(scope="module")
+def mattermost_image(request):
+    """Get the mattermost image name from the --mattermost-image argument.
+
+    Return a the mattermost image name
+    """
+    return request.config.getoption("--mattermost-image")
+
+
+@pytest_asyncio.fixture(scope="module")
+async def app(
+    ops_test: OpsTest,
+    app_name: str,
+    mattermost_image: str,
+):
+    """Mattermost charm used for integration testing.
+
+    Builds the charm and deploys it and the relations it depends on.
+    """
+    assert ops_test.model
+    await ops_test.model.deploy("postgresql-k8s"),
+
+    charm = await ops_test.build_charm(".")
+    application = await ops_test.model.deploy(charm, application_name=app_name, series="focal")
+    await ops_test.model.wait_for_idle()
+
+    # change the image that will be used for the mattermost container
+    await application.set_config(
+        {
+            "mattermost_image_path": mattermost_image,
+        }
+    )
+    await ops_test.model.wait_for_idle()
+
+    await asyncio.gather(
+        ops_test.model.add_relation(app_name, "postgresql-k8s:db"),
+    )
+    # mypy doesn't see that ActiveStatus has a name
+    await ops_test.model.wait_for_idle(status=ActiveStatus.name, raise_on_error=False)  # type: ignore
+
+    yield application
+
+
+@fixture(scope="module")
+def localstack_ip(request):
+    """Get the localstack IP from the --localstack-ip argument.
+
+    Return the IP address of a localstack instance
+    """
+    return request.config.getoption("--localstack-ip")
+
+
+@fixture(scope="module")
+def localstack_s3_config(localstack_ip: str) -> dict:
+    """Generate a s3_config dict.
+
+    Return a dict of s3 configurations
+    """
+    s3_config: dict = {
+        # Localstack doesn't require any specific value there, any random string will work
+        "credentials": {
+            "access-key": "my-lovely-key",
+            "secret-key": "this-is-very-secret",
```

Comment:
Should we use https://docs.python.org/3/library/secrets.html#secrets.token_hex for this?

Reply:
I used secrets for this but I don't think it was really necessary since localstack takes any secret key and this one isn't reused (and it's a test)

Reply:
it's just better not to get into the habbit of using something other that `secrets` to create token unless it is really needed, there is very little drawback